File: batou-rules/testdata/fixtures/zig/vulnerable/sql_injection.zig
const std = @import("std");
const sqlite = @import("sqlite");

// SQL injection via string concatenation into exec
pub fn findUser(db: *sqlite.Db, allocator: std.mem.Allocator, username: []const u8) !void {
    const query = try std.fmt.allocPrint(allocator, "SELECT * FROM users WHERE name = '{s}'", .{username});
    defer allocator.free(query);
    _ = try db.exec(query, .{});
}

// SQL injection via dynamic prepare
pub fn deleteRecord(db: *sqlite.Db, allocator: std.mem.Allocator, table: []const u8) !void {
    const query = try std.fmt.allocPrint(allocator, "DELETE FROM {s} WHERE id = 1", .{table});
    defer allocator.free(query);
    var stmt = try db.prepare(query);
    defer stmt.deinit();
}

File: batou-rules/testdata/fixtures/rust/vulnerable/sql_injection.rs
use sqlx;

// RS-003: SQL injection via format! macro
pub async fn find_user(pool: &sqlx::PgPool, username: &str) -> Result<(), sqlx::Error> {
    let query = format!("SELECT * FROM users WHERE username = '{}'", username);
    sqlx::query(&query).execute(pool).await?;
    Ok(())
}

// RS-003: SQL injection via string concatenation
pub async fn search_products(pool: &sqlx::PgPool, term: &str) -> Result<(), sqlx::Error> {
    let query = String::from("SELECT * FROM products WHERE name LIKE '%") + term + "%'";
    sqlx::query(&query.as_str()).execute(pool).await?;
    Ok(())
}

// RS-003: diesel::sql_query with format!
pub fn delete_record(conn: &mut PgConnection, id: &str) {
    let query = format!("DELETE FROM records WHERE id = {}", id);
    diesel::sql_query(&query).execute(conn).unwrap();
}
